File: backend/schemas/task.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime

# Import the Enum from the models to reuse it
from ..models.task import TaskStatus


# Shared properties
class TaskBase(BaseModel):
    title: str = Field(..., max_length=255)
    description: Optional[str] = None
    due_date: datetime
    status: TaskStatus = TaskStatus.PENDING

# Properties to receive via API on creation
class TaskCreate(TaskBase):
    # All fields from TaskBase are required for creation,
    # except description which is optional.
    # Status defaults to PENDING if not provided.
    pass

# Properties to receive via API on status update
class TaskUpdateStatus(BaseModel):
    status: TaskStatus

# Properties shared by models stored in DB
# (Currently same as TaskBase + id, but could differ later)
class TaskInDBBase(TaskBase):
    id: int

    model_config = {
        "from_attributes": True
    }

# Properties to return to client
class Task(TaskInDBBase):
    # Inherits id, title, description, due_date, status from TaskInDBBase
    pass

# Properties stored in DB
# (Currently same as Task, but could differ if we add sensitive fields)
class TaskInDB(TaskInDBBase):
    pass 

File: frontend/src/types/task.ts
// Matches the TaskStatus enum in the backend (backend/models/task.py)
export enum TaskStatus {
  PENDING = "PENDING",
  IN_PROGRESS = "IN_PROGRESS",
  COMPLETED = "COMPLETED",
}

// Matches the Task schema returned by the API (backend/schemas/task.py -> Task)
export interface Task {
  id: number;
  title: string;
  description?: string | null; // Optional in backend, allow null from DB
  due_date: string; // Represent datetimes as strings in TS/JSON
  status: TaskStatus;
}

// Matches the TaskCreate schema used for creating tasks (backend/schemas/task.py -> TaskCreate)
export interface TaskCreate {
  title: string;
  description?: string | null;
  due_date: string; // Use string for input, backend will parse
  status?: TaskStatus; // Optional, defaults to PENDING in backend
}

// Matches the TaskUpdateStatus schema (backend/schemas/task.py -> TaskUpdateStatus)
export interface TaskUpdateStatus {
  status: TaskStatus;
}
